cocoa/src/toga_cocoa/window.py

```diff
@@ -288,11 +354,98 @@ def get_visible(self):
     # Window state
     ######################################################################
 
-    def set_full_screen(self, is_full_screen):
-        current_state = bool(self.native.styleMask & NSWindowStyleMask.FullScreen)
-        if is_full_screen != current_state:
+    def get_window_state(self):
+        if bool(self.container.native.isInFullScreenMode()):
+            return WindowState.PRESENTATION
+        elif bool(self.native.styleMask & NSWindowStyleMask.FullScreen):
+            return WindowState.FULLSCREEN
+        elif bool(self.native.isZoomed):
+            return WindowState.MAXIMIZED
+        elif bool(self.native.isMiniaturized):
+            return WindowState.MINIMIZED
+        else:
+            return WindowState.NORMAL
+
+    def set_window_state(self, state):
+        # Since the requests to the OS for changing window states are non-blocking,
+        # if we are in the middle of processing a state, we need to store the
+        # user-requested state and apply the state when we have completed processing
+        # a transition. There are 2 types of callbacks:
+        # * EnteredState:
+        #   Here, we need to check if the current state is the same as the pending
+        #   state.
+        #   -- If yes: Clear the pending state variable and return.
+        #   -- If no: Apply NORMAL state, which will later apply the pending state
+        #             when the state is NORMAL.
+        # * ExitedState:
+        #   Here, since we are in NORMAL state, we just apply the pending state.
+        #   When we enter the user-requested pending state, then clear the pending
+        #   state variable and return.
+
+        if self._pending_state_transition:
+            self._pending_state_transition = state
+        else:
+            # If the app is in presentation mode, but this window isn't, then
+            # exit app presentation mode before setting the requested state.
+            if any(
+                window.state == WindowState.PRESENTATION and window != self.interface
+                for window in self.interface.app.windows
+            ):
+                self.interface.app.exit_presentation_mode()
+
+            current_state = self.get_window_state()
+            if current_state == state:
+                return
+            self._pending_state_transition = state
```

Comment:
`pending` here means that the previous state change request is not completed and that any new state requests should be put on `_pending_state_transition` instead of being applied directly. For example,
* Initially a state of `WindowState.MINIMIZED` is requested and we will not set `self._pending_state_transition`. 
https://github.com/beeware/toga/blob/2af1509bec66263cc392d74585a0b4df539a9f2c/cocoa/src/toga_cocoa/window.py#L397-L398
* Now `self._apply_state(WindowState.MINIMIZED)` will be called, which will call `self.native.setIsMiniaturized(True)`.
https://github.com/beeware/toga/blob/2af1509bec66263cc392d74585a0b4df539a9f2c/cocoa/src/toga_cocoa/window.py#L415-L416
* Since, `setIsMiniaturized()` is a non-blocking call, it will return immediately.

* Now a new state request of `WindowState.FULLSCREEN` is requested at `set_window_state()`, while the previous request to change state to `MINIMIZED` is not completed.
* But since, `self._pending_state_transition` is not set, it will not wait for the previous state change request of `MINIMIZED` to complete, and would apply `WindowState.NORMAL` directly. Hence, causing glitches.
https://github.com/beeware/toga/blob/2af1509bec66263cc392d74585a0b4df539a9f2c/cocoa/src/toga_cocoa/window.py#L395-L396

Since in both cases, `self._pending_state_transition` is required to be set for proper working, so it is set unconditionally here.
https://github.com/beeware/toga/blob/2af1509bec66263cc392d74585a0b4df539a9f2c/cocoa/src/toga_cocoa/window.py#L394-L398